queries/kotlin/highlights.scm

```diff
@@ -0,0 +1,186 @@
+; accessed field in blocks. `logger` in `logger.info("")`
+(statements (call_expression (navigation_expression (simple_identifier) @variable)))
+(indexing_expression (call_expression (navigation_expression (simple_identifier) @variable)))
+; `classProp` in `classProp.lastIndex`
+(statements (navigation_expression (simple_identifier) @variable))
+; `variable` in `variable = car.doors`
+(directly_assignable_expression (simple_identifier) @variable)
+
+
+;; Constants
+; Assume all-caps names are constants
+((simple_identifier) @constant
+ (#match? @constant "^[A-Z][A-Z_]+"))
+((interpolated_identifier) @constant
+ (#vim-match? @constant "^[A-Z][A-Z_]+"))
+
+
+(lambda_parameters) @parameter
+
+
+;; Builtin functions
+((simple_identifier) @function.macro
+ (#vim-match? @function.macro "^(commonPrefixWith|commonSuffixWith|endsWith|findAnyOf|findLastAnyOf|hasSurrogatePairAt|ifBlank|ifEmpty|indexOf|indexOfAny|isEmpty|isNotBlank|isNotEmpty|isNullOrBlank|isNullOrEmpty|lastIndexOf|lastIndexOfAny|lineSequence|lines|orEmpty|padEnd|padStart|removePrefix|removeRange|removeSuffix|removeSurrounding|replace|replaceAfter|replaceAfterLast|replaceBefore|replaceBeforeLast|replaceFirst|replaceRange|split|splitToSequence|startsWith|subSequence|substring|substringAfter|substringAfterLast|substringBefore|substringBeforeLast|trim|trimEnd|trimStart|containsKey|containsValue|filter|filterKeys|filterNot|filterNotTo|filterTo|filterValues|getOrElse|getOrPut|getValue|ifEmpty|isNotEmpty|isNullOrEmpty|mapKeys|mapKeysTo|mapValues|mapValuesTo|orEmpty|putAll|remove|toMap|toMutableMap|toPair|also|apply|let|run|takeIf|takeUnless|prependIndent|replaceIndent|replaceIndentByMargin|trimIndent|trimMargin|all|any|asIterable|asSequence|associate|associateBy|associateByTo|associateTo|associateWith|associateWithTo|chunked|chunkedSequence|count|drop|dropLast|dropLastWhile|dropWhile|elementAtOrElse|elementAtOrNull|filter|filterIndexed|filterIndexedTo|filterNot|filterNotTo|filterTo|find|findLast|first|firstOrNull|flatMap|flatMapIndexed|flatMapIndexedTo|flatMapTo|fold|foldIndexed|foldRight|foldRightIndexed|forEach|forEachIndexed|getOrElse|getOrNull|groupBy|groupByTo|groupingBy|indexOfFirst|indexOfLast|last|lastOrNull|map|mapIndexed|mapIndexedNotNull|mapIndexedNotNullTo|mapIndexedTo|mapNotNull|mapNotNullTo|mapTo|max|maxBy|maxByOrNull|maxOf|maxOfOrNull|maxOfWith|maxOfWithOrNull|maxOrNull|maxWith|maxWithOrNull|min|minBy|minByOrNull|minOf|minOfOrNull|minOfWith|minOfWithOrNull|minOrNull|minWith|minWithOrNull|none|onEach|onEachIndexed|partition|random|randomOrNull|reduce|reduceIndexed|reduceIndexedOrNull|reduceOrNull|reduceRight|reduceRightIndexed|reduceRightIndexedOrNull|reduceRightOrNull|reversed|runningFold|runningFoldIndexed|runningReduce|runningReduceIndexed|scan|scanIndexed|scanReduce|scanReduceIndexed|single|singleOrNull|slice|sumBy|sumByDouble|sumOf|take|takeLast|takeLastWhile|takeWhile|toCollection|toHashSet|toList|toMutableList|toSet|windowed|windowedSequence|withIndex|zip|zipWithNext)$"))
```

Comment:
Is there any specific difference between this category and the latter (with print, println)? They are not real marcos, right?

Reply:
correct! the difference is that these are builtin extention functions; like `someString.trim()` where .trim() must be called on a string, whilst the other category are standalone builtin functions that can be called anywhere. like `println(...)` or `val a = listOf(1, 2, 3)`.

the reason I divided them is that that is what Intellj does, and that is what I'm used to.

Reply:
so it's not really macros like in C. if this is to far a stretch, let me know and I will change it

Reply:
If you like the highlighting, you could add a new category `@method.builtin` and default link it to a highlight of your choice. You'd need to add it to CONTRIBUTING.md to make the linter happy.

Reply:
@tormodatt I've seen a few advent of code solutions for kotlin. I guess it's really a bit macro-like. So you could also leave it this way. Would you consider the PR as ready to merge?

Reply:
yeah, one could argue that this is kotlin answer to macros, but method.builtin is definetly correct as well.

maybe we can merge this now, and consider adding a category later.

Reply:
![image](https://user-images.githubusercontent.com/7189118/101282646-9824a700-37d6-11eb-8ee3-83d04c98e256.png)

Maybe fix the constant regex later. I guess you need to add `^...$` or else it will be a constant with two uppercase letter only.
 
Your highlight for the built-in methods is really handy.

Reply:
eksisting is `(#vim-match? @constant "^[A-Z][A-Z_]+"))`, the `+` makes it match more than two letters. I agree that we should add `$`, but then we need to match numbers as well. I'm not able to make that work though. Tried both:
`^[A-Z][A-Z_\\d]+$` and `^[A-Z][A-Z_\d]+$` but none of those matches `PLATE2` in `private const val PLATE2 = "asdf$PLATE"`. 

what regex rules should work here?

Reply:
> Your highlight for the built-in methods is really handy.

thanks! I made a simple script for fetching the functions from kotlin stdlib (just the most used classes for now). are there any suitable place to put this, or should I just put it in my own repo for now?

Reply:
I would just copy the regex from Python, C++ or Java for constants.
E.g. in c++ `((identifier) @constant
                        (#match? @constant "^[A-Z][A-Z_0-9]*$"))`

Reply:
It's possible that `\` do not completely work the way the should in match. they are fed like that into `vim.fn.match`

`vim.fn.match(vim.fn.escape('\\v'..your_text, '\\'))`